File: connectors/javascript/tests/integration/core/C08_Disconnection.test.ts
/**
 * C-08: Disconnection Tests
 *
 * Tests that the Connector can properly disconnect and that
 * the state is correct after disconnection.
 */

import { describe, test, expect, beforeEach, afterEach } from 'vitest';
import { BaseIntegrationTest } from '../helpers/BaseIntegrationTest.js';
import { Packet } from '../../../src/packet.js';

describe('C-08: Disconnection', () => {
    let testContext: BaseIntegrationTest;

    beforeEach(async () => {
        testContext = new BaseIntegrationTest();
        await testContext['beforeEach']();
    });

    afterEach(async () => {
        await testContext['afterEach']();
    });

    test('C-08-01: Disconnect closes the connection', async () => {
        // Given: Connected and authenticated
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('disconnectUser');

        expect(testContext['connector']!.isConnected).toBe(true);
        expect(testContext['connector']!.isAuthenticated).toBe(true);

        // When: Disconnect
        testContext['connector']!.disconnect();
        await testContext['delay'](500); // Wait for disconnection to complete

        // Then: Should be disconnected
        expect(testContext['connector']!.isConnected).toBe(false);
        expect(testContext['connector']!.isAuthenticated).toBe(false);
    });

    test('C-08-02: OnDisconnect event does not trigger when client disconnects', async () => {
        // Given: Connected and authenticated
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('clientDisconnectUser');

        let disconnectEventTriggered = false;
        testContext['connector']!.onDisconnect = () => {
            disconnectEventTriggered = true;
        };

        // When: Client disconnects intentionally
        testContext['connector']!.disconnect();
        await testContext['delay'](1000); // Wait for event

        // Then: OnDisconnect event should not be triggered (intentional disconnect)
        expect(disconnectEventTriggered).toBe(false);
    });

    test('C-08-03: Send fails after disconnect', async () => {
        // Given: Connected, authenticated, then disconnected
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('sendAfterDisconnectUser');
        testContext['connector']!.disconnect();
        await testContext['delay'](500);

        let receivedErrorCode: number | undefined;
        testContext['connector']!.onError = (_stageId, errorCode, _message) => {
            receivedErrorCode = errorCode;
        };

        // When: Try to send after disconnection
        const echoRequest = {
            content: 'Test',
            sequence: 1
        };
        const packet = Packet.create('EchoRequest', echoRequest);
        testContext['connector']!.send(packet);

        // Process pending callbacks
        testContext['connector']!.mainThreadAction();

        // Then: Disconnected error should occur
        // ErrorCode.Disconnected = 2
        expect(receivedErrorCode).toBe(2);
    });

    test('C-08-04: RequestAsync throws exception after disconnect', async () => {
        // Given: Connected, authenticated, then disconnected
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('requestAfterDisconnectUser');
        testContext['connector']!.disconnect();
        await testContext['delay'](500);

        // When: Try to request after disconnection
        const echoRequest = {
            content: 'Test',
            sequence: 1
        };
        const packet = Packet.create('EchoRequest', echoRequest);

        // Then: Should throw exception
        await expect(
            testContext['connector']!.request(packet)
        ).rejects.toThrow('Not connected');
    });

    test('C-08-05: Can reconnect after disconnect', async () => {
        // Given: First connection and disconnect
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('reconnectUser1');
        testContext['connector']!.disconnect();
        await testContext['delay'](500);

        expect(testContext['connector']!.isConnected).toBe(false);

        // When: Reconnect to new stage
        const reconnected = await testContext['createStageAndConnect']();

        // Then: Reconnection should succeed
        expect(reconnected).toBe(true);
        expect(testContext['connector']!.isConnected).toBe(true);

        // Authentication should be possible
        const authReply = await testContext['authenticate']('reconnectUser2');
        expect(authReply).toBeDefined();
    });

    test('C-08-06: Multiple disconnect calls are safe', async () => {
        // Given: Connected
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('multiDisconnectUser');

        // When: Call disconnect multiple times
        testContext['connector']!.disconnect();
        await testContext['delay'](200);

        // Should not throw exception
        expect(() => {
            testContext['connector']!.disconnect();
            testContext['connector']!.disconnect();
            testContext['connector']!.disconnect();
        }).not.toThrow();

        // Then: Should be disconnected
        expect(testContext['connector']!.isConnected).toBe(false);
    });

    test('C-08-07: Connector cleanup disconnects automatically', async () => {
        // Given: Create and connect new connector
        const tempConnector = new Connector();
        tempConnector.init({ requestTimeoutMs: 5000, heartbeatIntervalMs: 10000 });

        const stageInfo = await testContext['testServer'].createStage('TestStage');
        const wsUrl = testContext['testServer'].wsUrl;
        await tempConnector.connect(wsUrl, stageInfo.stageId);

        const authSuccess = await tempConnector.authenticate('TestStage', 'disposeUser');
        expect(authSuccess).toBe(true);
        expect(tempConnector.isConnected).toBe(true);

        // When: Disconnect (cleanup)
        tempConnector.disconnect();
        await testContext['delay'](500);

        // Then: Should be disconnected
        expect(tempConnector.isConnected).toBe(false);
    });

    test('C-08-08: IsAuthenticated returns false after disconnect', async () => {
        // Given: Connected and authenticated
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('authCheckUser');

        expect(testContext['connector']!.isAuthenticated).toBe(true);

        // When: Disconnect
        testContext['connector']!.disconnect();
        await testContext['delay'](500);

        // Then: IsAuthenticated should be false
        expect(testContext['connector']!.isAuthenticated).toBe(false);
    });
});

File: connectors/cpp/tests/integration/core/c08_disconnection_test.cpp
#include "../base_integration_test.hpp"
#include <gtest/gtest.h>
#include <atomic>

using namespace playhouse;
using namespace playhouse::test;

/// C-08: Disconnection Tests
/// Verifies proper disconnection handling
class C08_DisconnectionTest : public BaseIntegrationTest {};

TEST_F(C08_DisconnectionTest, Disconnect_AfterConnection_IsConnectedReturnsFalse) {
    // Given: Connected to server
    ASSERT_TRUE(CreateStageAndConnect());
    EXPECT_TRUE(connector_->IsConnected());

    // When: Disconnect
    connector_->Disconnect();
    std::this_thread::sleep_for(std::chrono::milliseconds(500));

    // Then: IsConnected should return false
    EXPECT_FALSE(connector_->IsConnected()) << "IsConnected should be false after disconnect";
}

TEST_F(C08_DisconnectionTest, Disconnect_OnDisconnectEvent_Fires) {
    // Given: Connected to server
    ASSERT_TRUE(CreateStageAndConnect());

    std::atomic<bool> disconnect_event_fired{false};

    connector_->OnDisconnect = [&]() {
        disconnect_event_fired = true;
    };

    // When: Disconnect
    connector_->Disconnect();

    // Wait for event with MainThreadAction
    bool completed = WaitForConditionWithMainThreadAction([&]() {
        return disconnect_event_fired.load();
    }, 5000);

    // Then: OnDisconnect event should fire
    EXPECT_TRUE(completed) << "OnDisconnect event should fire within timeout";
    EXPECT_TRUE(disconnect_event_fired) << "OnDisconnect event should be triggered";
}

TEST_F(C08_DisconnectionTest, Disconnect_MultipleTimesSafely_NoErrors) {
    // Given: Connected to server
    ASSERT_TRUE(CreateStageAndConnect());

    // When: Disconnect multiple times
    connector_->Disconnect();
    std::this_thread::sleep_for(std::chrono::milliseconds(200));

    connector_->Disconnect();  // Second disconnect
    std::this_thread::sleep_for(std::chrono::milliseconds(200));

    connector_->Disconnect();  // Third disconnect

    // Then: Should handle gracefully without errors
    EXPECT_FALSE(connector_->IsConnected());
    SUCCEED() << "Multiple disconnects handled safely";
}

TEST_F(C08_DisconnectionTest, Disconnect_BeforeConnection_HandledGracefully) {
    // Given: Not yet connected
    EXPECT_FALSE(connector_->IsConnected());

    // When: Try to disconnect
    connector_->Disconnect();

    // Then: Should handle gracefully
    EXPECT_FALSE(connector_->IsConnected());
    SUCCEED() << "Disconnect before connection handled safely";
}

TEST_F(C08_DisconnectionTest, Send_AfterDisconnection_TriggersError) {
    // Given: Connected then disconnected
    ASSERT_TRUE(CreateStageAndConnect());
    connector_->Disconnect();
    std::this_thread::sleep_for(std::chrono::milliseconds(500));

    std::atomic<bool> error_triggered{false};
    int error_code = 0;

    connector_->OnError = [&](int code, std::string message) {
        error_triggered = true;
        error_code = code;
    };

    // When: Try to send after disconnection
    std::string data = "{\"test\":\"data\"}";
    Bytes payload(data.begin(), data.end());
    connector_->Send(Packet::FromBytes("TestMessage", std::move(payload)));

    // Process callbacks
    for (int i = 0; i < 10; i++) {
        connector_->MainThreadAction();
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

    // Then: Error should be triggered
    EXPECT_TRUE(error_triggered) << "Error event should be triggered for send after disconnect";
    EXPECT_EQ(error_code, error_code::CONNECTION_CLOSED) << "Error code should indicate connection closed";
}

TEST_F(C08_DisconnectionTest, Reconnect_AfterDisconnection_Succeeds) {
    // Given: Connected then disconnected
    ASSERT_TRUE(CreateStageAndConnect());
    connector_->Disconnect();
    std::this_thread::sleep_for(std::chrono::milliseconds(500));

    // When: Reconnect
    auto new_stage = GetTestServer().GetOrCreateTestStage();
    (void)new_stage;
    bool reconnected = ConnectAndWait(5000);

    // Then: Reconnection should succeed
    EXPECT_TRUE(reconnected) << "Reconnection should succeed";
    EXPECT_TRUE(connector_->IsConnected()) << "Should be connected after reconnection";
}
